### ISSUE TITLE
std.lib: proposal: stringify: Print union(enum) field name

### ISSUE THREAD

**Issue by jdmichaud** [ziglang/zig#9735]
### Problem statement
As of today, `std.json.stringify` does not include the name of the field in the output:
```zig
const std = @import("std");

const Foo = struct { foo: u8 };
const Bar = struct { bar: u8 };

const T = union(enum) {
  A: Foo,
  B: Bar,
};

pub fn main() !void {
  try std.json.stringify(T { .A = .{ .foo = 42 } }, std.json.StringifyOptions {}, std.io.getStdOut().writer());
}
```
yields:
```zig
{"foo":42}
```

This is intentional to allow [arrays with different value types](https://github.com/ziglang/zig/issues/8913) as authorized in the json specs. However, it is then impossible to distinguish between two different types which would have the same structure:
```zig
[...]
const Foo = struct {};
const Bar = struct {};
[...]
```

Would yield:
```zig
{}
```
Which would make it impossible for the json parser to distinguish between the A or B type.

### Proposed solution
Add an option in [StringifyOptions](https://github.com/ziglang/zig/blob/a2ff3a13fe22b5693bbd719bf4eaee06a9d6be57/lib/std/json.zig#L2811) so that enum fields are named in the output, yielding:
```zig
{"A":{"foo":42}}
```

This would mean adding a condition [here](https://github.com/ziglang/zig/blob/a2ff3a13fe22b5693bbd719bf4eaee06a9d6be57/lib/std/json.zig#L2932) to add the field name.

### Use case
The general use case is to lift conditions on `union(enum)` for them to be serializable in json. Right now, the above conditions apply (each field must be a distinguishable json format).

A particular case is the use of an `union(enum)` as a "protocol" between a client and a server. Each field of the enum is a different message. Some message have data and some don't. Some message have the same data. It is very practical, on the receiving end, to be able to parse the enum and `switch()` on it. But if some fields have no data or identical data, parsing fails.

### Alternative
An alternative was proposed to use a struct with optional fields, but that would make things weird and not as elegant as the enum solution.

### Workaround
In case this proposal is rejected, a possible workaround is to add a dummy field, unique to each field:
```zig
[...]
const Foo = struct { __foo__: u8 };
const Bar = struct { __bar__: u8 };
[...]
```
then:
```zig
  try std.json.stringify(T { .A = .{ .__foo__ = 0 } }, std.json.StringifyOptions {}, std.io.getStdOut().writer());
```

### Rebuttal

The argument against this proposal would be that `std.io.json` is not a serialization library but a library to easily allow the creation and parsing of valid json. In that case, representing all Zig data structures in json is not the priority. The priority is to allow all possible json construct.

**Comment by daurnimator** [comment]
The intended way to get the serialisation you're after is:

```zig
const T = union(enum) {
  A: struct { A: Foo },
  B: struct { B: Bar},
};
```

**Comment by jdmichaud** [comment]
@daurnimator yes it is equivalent of having a specific field in your original struct as in the workaround I proposed. But you have to change your intended structure to explicitly distinguish the different fields.

**Comment by daurnimator** [comment]
> But you have to change your intended structure to explicitly distinguish the different fields.

What do you mean?

**Comment by jdmichaud** [comment]
I mean that it all boils down to what `stringify` is trying to accomplish. If is is the equivalent of `JSON.stringify` from javascript, then what we want is to serialize Zig data structures (with the the objective of deserializing them at some point). We should probably not have to modify the structure so that they can be serialized/deserialized properly.

**Comment by daurnimator** [comment]
> If is is the equivalent of `JSON.stringify` from javascript, then what we want is to serialize Zig data structures (with the the objective of deserializing them at some point).

The goal was never to stringify arbitrary zig types; but yet `serialize` is intended to be a mirror of `parse`. 

----



Note that you *can* write e.g.:
```zig
const MyStruct = union(enum) {
  A: Foo,
  B: Bar,
  pub fn jsonStringify(
      value: @This(),
      options: json.StringifyOptions,
      out_stream: anytype,
    ) @TypeOf(out_stream).Error!void { 
    const jsonT = union(std.meta.Tag(@This())) {
      A: struct { A: Foo },
      B: struct { B: Bar },
    }
    return json.stringify(@as(jsonT, switch(value) {
      .A => |a| .{ .A = a },
      .B => |b| .{ .B = b },
    }), options, out_stream);
  }
};
```

**Comment by jdmichaud** [comment]
For reference, this is [what `serde` does](https://serde.rs/enum-representations.html). It helps put name on things. What zig's stringify does is `Untagged` serialization of enums.
